sonar-server/atomic-server.go

```diff
@@ -0,0 +1,211 @@
+	eID := atomic.AddInt32(&s.eventID, 1)
+	ew := eventWrapper{
+		eventID:         eID,
+		eventType:       request.OperationType,
+		eventObject:     request.Object,
+		eventObjectType: request.ResourceType,
+	}
+	log.Println("NotifyAtomicBeforeIndexerWrite", ew.eventID, ew.eventType, ew.eventObjectType, ew.eventObject)
+	currentEvent := strToMap(request.Object)
+	crucialCurEvent := strToMap(s.crucialCur)
+	crucialPrevEvent := strToMap(s.crucialPrev)
+	// We then check for the crucial event
+	if ew.eventObjectType == s.ceRtype && getEventResourceName(currentEvent) == s.ceName && getEventResourceNamespace(currentEvent) == s.ceNamespace {
+		log.Print("[sonar] we then check for crash condition", "s.crash", s.crash, "s.seenPrev", s.seenPrev)
+		if s.shouldCrash(crucialCurEvent, crucialPrevEvent, currentEvent) {
+			log.Println("[sonar] should crash the operator while issuing target side effect")
+			s.crucialEvent = ew
+		}
+	}
+	*response = sonar.Response{Message: request.OperationType, Ok: true, Number: int(eID)}
+	return nil
+}
+
+func (s *atomicServer) NotifyAtomicSideEffects(request *sonar.NotifyAtomicSideEffectsRequest, response *sonar.Response) error {
+	name, namespace := extractNameNamespace(request.Object)
+	log.Printf("[SONAR-SIDE-EFFECT]\t%s\t%s\t%s\t%s\t%s\n", request.SideEffectType, request.ResourceType, namespace, name, request.Error)
+	if s.crash && !s.restarted && request.ResourceType == s.seRtype && request.SideEffectType == s.seEtype && name == s.seName && namespace == s.seNamespace {
+		// we should restart operator here
+		s.restarted = true
+		log.Printf("we restart operator pod here\n")
+		s.restartComponent()
+		return nil
+	}
+	*response = sonar.Response{Message: request.SideEffectType, Ok: true}
+	return nil
+}
+
+func (s *atomicServer) restartComponent() {
+	masterUrl := "https://" + s.frontRunner + ":6443"
+	config, err := clientcmd.BuildConfigFromFlags(masterUrl, "/root/.kube/config")
+	checkError(err)
+	clientset, err := kubernetes.NewForConfig(config)
+	checkError(err)
+
+	deployment, err := clientset.AppsV1().Deployments(s.namespace).Get(context.TODO(), s.deployName, metav1.GetOptions{})
+	checkError(err)
+	log.Println(deployment.Spec.Template.Spec.Containers[0].Env)
+
+	clientset.AppsV1().Deployments(s.namespace).Delete(context.TODO(), s.deployName, metav1.DeleteOptions{})
+
+	labelSelector := metav1.LabelSelector{MatchLabels: map[string]string{"sonartag": s.podLabel}}
+	listOptions := metav1.ListOptions{
+		LabelSelector: labels.Set(labelSelector.MatchLabels).String(),
+	}
+
+	for true {
+		time.Sleep(time.Duration(1) * time.Second)
+		pods, err := clientset.CoreV1().Pods(s.namespace).List(context.TODO(), listOptions)
+		checkError(err)
+		if len(pods.Items) != 0 {
+			log.Printf("operator pod not deleted yet\n")
+		} else {
+			log.Printf("operator pod gets deleted\n")
+			break
+		}
+	}
+
+	time.Sleep(time.Second * 15)
```

Comment:
As discussed yesterday, let's start with the most simple form without the 15s hardcoded sleep.